Trust proxy headers for host provision callback

chrismeyersfsu opened the pull request:
##### SUMMARY
* Do not remove special header list if request is from a trusted proxy.
* Continue to remove headers if request if from a non-trusted proxy.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
devel
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

chrismeyersfsu committed:
Trust proxy headers for host provision callback

* Do not remove special header list if request is from a trusted proxy.
* Continue to remove headers if request if from a non-trusted proxy.

chrismeyersfsu committed:
Tests for trust proxy and existing explicit proxy

* Integration tests to ensure the integration of the two features.

AlanCoding commented:
https://github.com/ansible/django-ansible-base/pull/454 and https://github.com/ansible/django-ansible-base/pull/452 are also very proxy-ish. What's the dependency chain of these patches?

chrismeyersfsu commented:
> [ansible/django-ansible-base#454](https://github.com/ansible/django-ansible-base/pull/454) and [ansible/django-ansible-base#452](https://github.com/ansible/django-ansible-base/pull/452) are also very proxy-ish. What's the dependency chain of these patches?

Quality of life improvements and they go towards “though shall not throw a 500”.

I just found them while writing tests in awx. The header should normally be well formed.

john-westcott-iv commented on awx/api/generics.py:
```diff
@@ -153,13 +154,21 @@ def initialize_request(self, request, *args, **kwargs):
         # If there are any custom headers in REMOTE_HOST_HEADERS, make sure
         # they respect the allowed proxy list
-        if all(
+        if not is_trusted_proxy and all(
```
I see you have tests below but I am reading this as if:
    * I am not trusted and
    * I have PROXY_IP_ALLOWED_LIST and
    * My REMOTE_ADDRS it not in PROXY_IP_ALLOWED_LIST and
    * My REMOTE_HOST is not in PROXY_IP_ALLOWED_LIST
then remove the headers.

Wouldn't the not_trusted just let you decide if you can look through more headers than just REMOTE_ADDRS and REMOTE_HOST and not whether or not we should remove headers?

chrismeyersfsu replied:
Yes. Making the necessary changes.

chrismeyersfsu committed:
Revert "Trust proxy headers for host provision callback"

This reverts commit 49e3971cd577127705fc0fd1d3b4ab7e3a3c3c2b.

chrismeyersfsu committed:
Add support for x-trusted-proxy

* Increase the surface area of the set of headers that the proxy list
  feature looks at for the remote proxy IF x-trusted-proxy is valid.

AlanCoding commented on awx/main/utils/proxy.py:
```diff
@@ -0,0 +1,30 @@
+"""
+
+
+def is_proxy_in_headers(request: Request, proxy_list: list[str], headers: list[str]) -> bool:
```
I'd like a little more documentation on what `proxy_list` is here, because its name doesn't match well with what it actually seems to be. It seems that it corresponds to the proxy allow list setting? So this is a list of IPs we will accept as proxies?

AlanCoding replied:
And why call it `headers` here and `remote_headers` in other places?

So this is like a list of the keys of headers, that may specify the IP that it got forwarded from?

chrismeyersfsu replied:
> I'd like a little more documentation on what proxy_list is here, because its name doesn't match well with what it actually seems to be. It seems that it corresponds to the proxy allow list setting? So this is a list of IPs we will accept as proxies?

A list of ips or hostnames. It's more than _likely_ ips. Some customer _might_ put a hostname to we would match against the special case `REMOTE_HOST` (described below).

```
Any security-related use of X-Forwarded-For (such as for rate limiting or IP-based access control) must only use IP addresses added by a trusted proxy. Using untrustworthy values can result in rate-limiter avoidance, access-control bypass, memory exhaustion, or other negative security or availability consequences.
```

> And why call it headers here and remote_headers in other places?

No reason. To add the confusion even more there are 2 special entries that actually aren't headers. `REMOTE_ADDR` and `REMOTE_HOST`. They are more like env vars, set by uwsgi, but they are in this `request.environ` dict. They don't start with `HTTP_`, so that distinguishes them in Django land.

> So this is like a list of the keys of headers, that may specify the IP that it got forwarded from?

yep

john-westcott-iv commented on awx/api/generics.py:
```diff
@@ -153,22 +155,23 @@ def initialize_request(self, request, *args, **kwargs):
 
+        if 'HTTP_X_TRUSTED_PROXY' in request.environ:
+            if validate_x_trusted_proxy_header(request.environ['HTTP_X_TRUSTED_PROXY']):
+                remote_headers = settings.REMOTE_HOST_HEADERS
```
Should this be
```suggestion
                remote_headers.extend(settings.REMOTE_HOST_HEADERS)
```

chrismeyersfsu replied:
I don't think so. `REMOTE_HOST_HEADERS` defaults to `['REMOTE_ADDR', 'REMOTE_HOST']`.

john-westcott-iv replied:
Ah, ok; that should be fine than. If a user took them out for some reason that would just be "on them".

john-westcott-iv commented on awx/main/utils/proxy.py:
```diff
@@ -0,0 +1,30 @@
+    return bool(remote_hosts.intersection(set(proxy_list)))
+
+
+def delete_headers(request: Request, headers: list[str]):
```
Should we maybe change this to `delete_headers_starting_with_http` or `delete_insecure_headers`?

chrismeyersfsu replied:
yes

john-westcott-iv commented on awx/api/generics.py:
```diff
@@ -153,22 +155,23 @@ def initialize_request(self, request, *args, **kwargs):
-                    request.environ.pop(custom_header, None)
+        if settings.PROXY_IP_ALLOWED_LIST:
+            if not is_proxy_in_headers(self.request, settings.PROXY_IP_ALLOWED_LIST, remote_headers):
+                delete_headers(request, settings.REMOTE_HOST_HEADERS)
```
I'm also wondering if we want to move this into DAB so that other services could just call something like:
```
from ansible_base.lib.utils.requests import delete_insecure_headers_if_not_from_trusted_proxy
delete_insecure_headers_if_not_from_trusted_proxy()
```

chrismeyersfsu replied:
I'm hoping this all goes away in the future and we use mtls and open Envoy up to be more configurable.

AWX shouldn't be deciding which headers are trusted and which aren't. That should be done by the proxies in front of it. You need to secure the entire chain of proxies and headers, doing it at the last point is not ideal.

If any of this stays, it should be moved to middleware.

chrismeyersfsu committed:
Rename delete

* Include a bit of context into the name of the delete function. The
  HTTP_ added prepended string may be unexpected if Django's header
  transformation isn't top of mind.

chrismeyersfsu committed:
Clarify the search for a proxy
